src/traits/wf.md

```diff
@@ -1,11 +1,440 @@
+}
+// The only non-parameter type which appears in this definition is
+// `<T as Iterator>::Item`. The generated goal is the following:
+// ```
+// forall<T> {
+//     WellFormed(<T as Iterator>::Item)
+// }
+// ```
+// which is not provable, hence the trait definition is considered illegal.
+```
+
+```rust,ignore
+trait Baz {
+    type Assoc<T>: From<<T as Iterator>::Item> where T: Iterator;
+}
+// The generated goal is now:
+// ```
+// forall<T> {
+//     if (FromEnv(T: Iterator)) {
+//         WellFormed(<T as Iterator>::Item)
+//     }
+// }
+// ```
+// which is now provable.
+```
+
+# Impls
+
+Now we give ourselves a general impl for the trait defined above:
+```rust,ignore
+impl<P1...> Trait<A1...> for SomeType<A2...> where WC_impl {
+    type Assoc<P2...> = SomeValue<A3...> where WC_assoc;
+}
+```
+
+Note that here, `WC_assoc` are the same where clauses as those defined on the
+associated type definition in the trait declaration, *except* that type
+parameters from the trait are substituted with values provided by the impl
+(see example below). You cannot add new where clauses. You may omit to write
+the where clauses if you want to emphasize the fact that you are actually not
+relying on them.
+
+Some examples to illustrate that:
+```rust,ignore
+trait Foo<T> {
+    type Assoc where T: Clone;
+}
+
+struct OnlyClone<T: Clone> { ... }
+
+impl<T> Foo<Option<T>> for () {
+    // We substitute type parameters from the trait by the ones provided
+    // by the impl, that is instead of having a `T: Clone` where clause,
+    // we have an `Option<T>: Clone` one.
+    type Assoc = OnlyClone<Option<T>> where Option<T>: Clone;
+}
+
+impl<T> Foo<T> for i32 {
+    // I'm not using the `T: Clone` where clause from the trait, so I can
+    // omit it.
+    type Assoc = u32;
+}
+
+impl<T> Foo<T> for f32 {
+    type Assoc = OnlyClone<Option<T>> where Option<T>: Clone;
+    //                                ^^^^^^^^^^^^^^^^^^^^^^
+    //                                this where clause does not exist
+    //                                on the original trait decl: illegal
+}
+```
+
+So where clauses on associated types work *exactly* like where clauses on
+trait methods: in an impl, we must substitute the parameters from the traits
+with values provided by the impl, we may omit them if we don't need them, and
+we cannot add new where clauses.
+
+Now let's see the generated goal for this general impl:
+```text
+forall<P1...> {
+    if (FromEnv(WC_impl), FromEnv(InputTypes(SomeType<A2...>))) {
+        WellFormed(SomeType<A2...>: Trait<A1...>) &&
+            WellFormed(InputTypes(WC_impl)) &&
+
+            forall<P2...> {
+                if (FromEnv(WC_assoc)) {
+                    WellFormed(SomeValue<A3...>: Bounds_assoc) &&
+                        WellFormed(InputTypes(SomeValue<A3...>))
+                }
+            }
+    }
+}
+```
+
+Here is the most complex goal. As always, a first thing is that assuming that
```

Comment:
```suggestion
Here is the most complex goal. As always, first, assuming that
```

Reply:
Also, what part of the goal does this correspond to? Is it `WellFormed(InputTypes(WC_impl))`? If so, could we reorder the two lines in the goal so they match the order discussed in the paragraphs.

Reply:
This does not really correspond to a part actually because it spreads all through the goal, since there are nested where clauses etc, i.e. when checking well-formedness of the input types of the associated type value, we can assume that the where clauses defined on the assoc type hold.

So this corresponds to both `WellFormed(InputTypes(WC_impl))` and `WellFormed(InputTypes(SomeValue<A3...>))` but we don't have the same assumptions in our environment for these two goals.

Not sure how to best format this, if you have any suggestions... :)